File: app/src/main/java/com/nytimes/android/external/register/RegisterApp.java
package com.nytimes.android.external.register;

import android.app.Application;
import android.content.Context;
import android.support.annotation.NonNull;

import com.nytimes.android.external.register.di.ApplicationComponent;
import com.nytimes.android.external.register.di.ApplicationModule;
import com.nytimes.android.external.register.di.DaggerApplicationComponent;
import com.nytimes.android.external.register.di.Injector;

public class RegisterApp extends Application {

    protected ApplicationComponent applicationComponent;
    protected ApplicationModule applicationModule;

    @NonNull
    public static RegisterApp get(@NonNull Context context) {
        return (RegisterApp) context.getApplicationContext();
    }

    protected void initDagger() {
        applicationModule = new ApplicationModule(this);
        applicationComponent = DaggerApplicationComponent.builder()
                .applicationModule(applicationModule)
                .build();
    }

    @Override
    public void onCreate() {
        super.onCreate();
        initDagger();
    }

    @Override
    public Object getSystemService(@NonNull String name) {
        if (Injector.matchesApp(name)) {
            return applicationComponent;
        }
        return super.getSystemService(name);
    }
}

File: app/src/main/java/com/nytimes/android/external/register/RegisterApp.kt
package com.nytimes.android.external.register

import android.app.Application
import com.nytimes.android.external.register.di.ApplicationComponent
import com.nytimes.android.external.register.di.ApplicationModule
import com.nytimes.android.external.register.di.DaggerApplicationComponent
import com.nytimes.android.external.register.di.Injector

class RegisterApp : Application() {

    private lateinit var applicationComponent: ApplicationComponent
    private lateinit var applicationModule: ApplicationModule

    override fun onCreate() {
        super.onCreate()
        initDagger()
    }

    private fun initDagger() {
        applicationModule = ApplicationModule(this)
        applicationComponent = DaggerApplicationComponent.builder()
                .applicationModule(applicationModule)
                .build()
    }

    override fun getSystemService(name: String): Any? {
        return if (Injector.matchesApp(name)) {
            applicationComponent
        } else super.getSystemService(name)
    }

}
